# Name : Samarth Deshpande
# Student ID  : 8894058

# Importing Libraries

In [9]:
import numpy as np
import pandas as pd
from sklearn.metrics import r2_score, mean_absolute_error,mean_squared_error,mean_absolute_percentage_error
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

# Loading Dataset

In [5]:
from sklearn import datasets
X, y = datasets.load_diabetes(return_X_y=True, as_frame=True , scaled=True)

# Splitting into indepndent and dependent variable

In [6]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [7]:
from sklearn.model_selection import cross_validate

# List to store degree from 0 to 9 and also to store R2, MAE, MAPE
degrees = []
r2_mean_scores = []
mae_mean_scores = []
mape_mean_scores = []
r2_std_scores = []
mae_std_scores = []
mape_std_scores = []

for degree in range(10):
    # Creating Polynomial Function
    poly_features = PolynomialFeatures(degree=degree)
    X_poly = poly_features.fit_transform(X_train)

    # Performing cross-validation using 10 splits
    model = LinearRegression()
    scores = cross_validate(model, X_poly, y_train, cv=10, n_jobs=-1, verbose=2, scoring=['r2', 'neg_mean_absolute_error', 'neg_mean_absolute_percentage_error'])

    degrees.append(degree)
    r2_mean_scores.append(np.mean(scores['test_r2']))
    mae_mean_scores.append(-np.mean(scores['test_neg_mean_absolute_error']))
    mape_mean_scores.append(-np.mean(scores['test_neg_mean_absolute_percentage_error']))
    r2_std_scores.append(np.std(scores['test_r2']))
    mae_std_scores.append(np.std(scores['test_neg_mean_absolute_error']))
    mape_std_scores.append(np.std(scores['test_neg_mean_absolute_percentage_error']))

results_df = pd.DataFrame({'Degree': degrees, 'R2_mean': r2_mean_scores, 'MAE_mean': mae_mean_scores, 'MAPE_mean': mape_mean_scores, 'R2_std': r2_std_scores, 'MAE_std': mae_std_scores, 'MAPE_std': mape_std_scores})

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   7 out of  10 | elapsed:    3.9s remaining:    1.6s
[Parallel(n_jobs=-1)]: Done  10 out of  10 | elapsed:    4.0s finished
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   7 out of  10 | elapsed:    0.6s remaining:    0.2s
[Parallel(n_jobs=-1)]: Done  10 out of  10 | elapsed:    0.6s finished
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   7 out of  10 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  10 out of  10 | elapsed:    0.0s finished
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   7 out of  10 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  10 out of  10 | elapsed:    0.0s finished
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_

In [8]:
results_df

Degree     R2_mean    MAE_mean  MAPE_mean      R2_std     MAE_std  MAPE_std
0       0   -0.070422   66.704968   0.631643    0.128025   10.247625  0.105315
1       1    0.393217   45.448783   0.405877    0.288324    3.861528  0.091139
2       2    0.235903   50.204970   0.432936    0.406517    5.484428  0.091591
3       3 -151.473209  339.743571   2.821005  243.889645  130.448712  1.236895
4       4  -62.021606  329.649942   2.756321   76.121104   99.917160  0.958981
5       5  -62.617869  321.930511   2.706063   85.348857  102.090830  0.989590
6       6  -62.627073  321.876398   2.705695   85.428768  102.106815  0.989760
7       7  -62.631675  321.873178   2.705688   85.446406  102.112017  0.989832
8       8  -62.632409  321.875584   2.705705   85.447207  102.111603  0.989825
9       9  -62.632453  321.874197   2.705693   85.448059  102.111771  0.989825

# My inferences on Cross-validation results

The R2 values are positve for degrees  1, and 2, which indicate a good fit. But the values decreases as the degree increases, which means  model shows poor fit.
Here Polynomial dgree 3 is worst because R2 score is high negativly, MAE and MAPE is also high.

# I would choose

# The Polynomial degree 1 because 
In comparison to the other models, it has the greatest R2 value 0.3932, meaning that it explains more variance in the target variable.
MAPE is also good 0.40 which is lowest, which shows that its predictions are the closest to the actual values.
MAE is also less so I would choose multi varient multi linear regression with polynomial degree with 1.